### DISCUSSION TITLE
can i detect clicked object stroke/fill?

### DISCUSSION THREAD

**Discussion by KiDeokKim** [fabricjs/fabric.js#7892] [Q&A] [answered]
Hi, i'm trying to develope extract object color & apply color (like drawing tool in windows)
i can saved color value when i clicked object or something, but i don't know that how to detect clicked object's fill or stroke.
i read docs and searched some articles, but i can't find solution.

how to detected click object's fill or stroke?
Example - Rectangles, circles, etc. have lines or fillings. I want to check whether it is the border or the inside of the object when I click it from the mouse.

pick color sample code : https://gist.github.com/RahulParmarRP/79fdd20f8c8aa47355f1ec66764d18d3

**Comment by ShaMan123** [comment, marked as answer]
Look at this method and adapt it to your needs
https://github.com/fabricjs/fabric.js/blob/cc363129b769a148bfe0599ffb46eeba3f3da640/src/canvas.class.js#L524

**Reply by KiDeokKim** [reply]
thank you for your answer. However, when the object has already set the fill color, it also returns false...
when the color of object is already set, is there no other function or detected outline / inside function?

**Reply by ShaMan123** [reply]
Of course it returns false, it's not transparent.
You need to adapt the method for your needs

**Comment by asturur** [comment]
@KiDeokKim this seems the xy problem.
You want to build a colorpicker, you started with an implementation, now you need to differentiate between stroke and fill color, and so you are asking how to understand how to detected a border click or a fill click.
The point is that you want a color picker.
You can access the canvas current color at mouse coordinates and detect any color clicked.

on the mousedown event you get a pointer with x and y coordinates.
You can access the canvas contextContainer and execute a getImageData there, you can do the getImageData at x,y,1,1 where 1 and 1 are width and height of a pixel.

https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getImageData

that will give you the color value udner the mouse at any moment.

**Reply by KiDeokKim** [reply]
thank you for your answer.
I think you are misunderstanding because my English writing skills are low, but I am already extracting the color value through the mouse.
(you are right : you need to differentiate between stroke and fill color, and so you are asking how to understand how to detected a border click or a fill click.)
I just wanted to ask if it was a border or a filling when I clicked on an object through a mouse.

**Reply by asturur** [reply]
wouldn't you want to know just what color was under the mouse? Like if the fill is a gradient, what color do you want to know?
Anyway, apart for RECTS and CIRCLES, where the math is simple, no there is no way to know if you clicked a fill or stroke, unless as @ShaMan123 suggested you get the `isTargetTransparent` and you modifiy to be `isTargetStroke` and `isTargetFill`. You would need to take in account paintfirst property too to know when fill and stroke overlap, which is the one on the top.

**Reply by ShaMan123** [reply]
I have another idea which might be an overkill.
You could have every object be replaced by 2 objects - one for the fill and one for the stroke and then use fabric to simply detect the click with default mechanism and listen to the `object:selected` event.

**Reply by ShaMan123** [reply]
group is ripe for that

**Reply by asturur** [reply]
just inspect the color under the mouse, and if is the same color of stroke, you know it was stroke, if is same color of fill, it was fill.
If is an aliased pixel, well you would have issues with the other method too.

**Reply by ShaMan123** [reply]
Another thought, setting having coords set with and without strokeWidth and then casting a ray from mouse pointer... that could work for polygons. But @asturur is right.

**Reply by KiDeokKim** [reply]
i don't consider for gradient color. i'm just consider for single color.
just think about windows paint application. in this, having tools pick the color next to the eraser tool and paint bucket tools above pick the color.
1. create some object and set stroke / fill color (another color)
2. extract color from pick the color tool ( i can extract/save 1 pixel color from mouse button event)
3. clicked paint bucket tool and when i clicked some object's stroke, stroke color is changed.

that's it. currently when i apply color from extracted color,  my application function just show message that 'apply **both** / **fill** / **stroke** color'

**Reply by ShaMan123** [reply]
This is exactly what @asturur suggested
